docs/source/user-guide/tuning.md

```diff
@@ -17,18 +17,96 @@ specific language governing permissions and limitations
 under the License.
 -->
 
-# Tuning Guide
+# Comet Tuning Guide
 
 Comet provides some tuning options to help you get the best performance from your queries.
 
 ## Memory Tuning
 
-### Unified Memory Management with Off-Heap Memory
+It is necessary to specify how much memory Comet can use in addition to memory already allocated to Spark. In some
+cases, it may be possible to reduce the amount of memory allocated to Spark so that overall memory allocation is
+the same or lower than the original configuration. In other cases, enabling Comet may require allocating more memory
+than before. See the [Determining How Much Memory to Allocate] section for more details.
 
-The recommended way to share memory between Spark and Comet is to set `spark.memory.offHeap.enabled=true`. This allows
-Comet to share an off-heap memory pool with Spark. The size of the pool is specified by `spark.memory.offHeap.size`. For more details about Spark off-heap memory mode, please refer to Spark documentation: https://spark.apache.org/docs/latest/configuration.html.
+[Determining How Much Memory to Allocate]: #determining-how-much-memory-to-allocate
 
-The type of pool can be specified with `spark.comet.exec.memoryPool`.
+Comet supports Spark's on-heap (the default) and off-heap mode for allocating memory. However, we strongly recommend
+using off-heap mode. Comet has some limitations when running in on-heap mode, such as requiring more memory overall,
+and requiring shuffle memory to be separately configured.
+
+### Configuring Comet Memory in Off-Heap Mode
+
+The recommended way to allocate memory for Comet is to set `spark.memory.offHeap.enabled=true`. This allows
+Comet to share an off-heap memory pool with Spark, reducing the overall memory overhead. The size of the pool is
+specified by `spark.memory.offHeap.size`. For more details about Spark off-heap memory mode, please refer to
+Spark documentation: https://spark.apache.org/docs/latest/configuration.html.
```

Comment:
Should we mention that `spark.comet.memoryOverhead` is used for off-heap as well? Otherwise, it will involve `spark.comet.memory.overhead.factor`?

Reply:
In `CometDriverPlugin`, we only reference this config if `shouldOverrideMemoryConf` returns `true`. The logic in that method is `cometEnabled && (cometExecShuffle || cometExec) && !unifiedMemory`.

`unifiedMemory` here really means `spark.memory.offHeap.enabled`.

The intent is not to use this config when running off-heap, because we define the overhead using `spark.memory.offHeap.size` when running in off-heap mode, although there may be places where we are not consistent. I will file an issue to check that we are using this consistently and have adequate tests.

Reply:
Not `CometDriverPlugin`, but I meant `getCometMemoryOverheadInMiB` and eventually this gets used in `CometExecIterator`

Reply:
It looks like this only gets used in off-heap mode if the default memory pool is changed to `fair_unified`. This seems incorrect to me since we should be using `spark.memory.offHeap.size` when in off-heap mode?

Reply:
Never mind. `getCometMemoryOverheadInMiB` uses `spark.memory.offHeap.size` when in off-heap mode, and does not use `spark.executor.memory`.

Reply:
It does still apply `COMET_MEMORY_OVERHEAD_FACTOR` though. This doesn't seem correct to me, but I will update the tuning guide to explain the current behavior and file a follow-on issue to discuss this more.

Reply:
Thanks, the memory limit for `fair_unified` is defined in a same way for the on-heap `fair_spill_global`.
I thought `spark.memory.offHeap.size` is just the amount that Spark recognize, and Spark needs some memory for off heap. I am thinking `spark.comet.memoryOverhead` more for avoiding to use all off-heap memory and leaving some off heap memory for Spark...

Reply:
> Thanks, the memory limit for `fair_unified` is defined in a same way for the on-heap `fair_spill_global`.

It is similar but not completely the same. `fair_unfied` will use `spark.memory.offHeap.size` as the base memory but  `fair_spill_global` will use `spark.executor.memory`. I pushed an update to specify how the pool size can be configured for `fair_unified`.

Reply:
I thought `spark.memory.offHeap.size` is just the amount that Spark recognize, and Spark needs some memory for off heap. I am thinking `spark.comet.memoryOverhead` more for avoiding to use all off-heap memory and leaving some off heap memory for Spark...

Ok, that is quite different from my understanding of what we were trying to achieve by introducing unified memory management so that Comet and Spark can use a unified memory pool to share off-heap memory. I think that is where some of the complexity has come in because the architecture of this is not documented well enough and I guess we have different understandings of what the design goal was.

@viirya Please add your thoughts if you have the time.

Reply:
Hmm, I took a look at current `getCometMemoryOverheadInMiB`. It mixes both unified memory and Spark on-heap memory mode cases.

Originally, it calculates off-heap memory required by Comet for Spark on-heap memory mode (when Comet doesn't provide unified memory mode). It is based on a factor (`COMET_MEMORY_OVERHEAD_FACTOR`) * Spark executor memory (on-heap). Or, you can specify it explicitly, i.e., `COMET_MEMORY_OVERHEAD` (i.e., `spark.comet.memoryOverhead`).

For the case using unified memory and sharing off-heap with Spark, I think we don't need to specify `COMET_MEMORY_OVERHEAD_FACTOR` or `COMET_MEMORY_OVERHEAD` anymore. It it because Comet should share all off-heap memory with Spark and the memory manager is the same (unified memory manager).

Reply:
Thanks @viirya. That is very helpful. I will go ahead and merge this PR since it generally improves the documentation but will follow up with another PR this week to try and address these points.